src/signatures.rs

```diff
@@ -0,0 +1,1206 @@
+///
+/// Account signatures should be thought of as a nested map, indexed on the
+/// outer layer by credential indices, and the inner map maps key indices to
+/// [`Signature`]s.
+#[derive(Debug, PartialEq, Eq, serde::Deserialize, serde::Serialize)]
+#[repr(transparent)]
+pub struct AccountSignatures {
+    pub sigs: BTreeMap<u8, CredentialSignatures>,
+}
+
+#[derive(Debug, PartialEq, Eq, serde::Deserialize, serde::Serialize)]
+#[repr(transparent)]
+pub struct CredentialSignatures {
+    pub sigs: BTreeMap<u8, Signature>,
+}
+
+impl AccountSignatures {
+    /// Create the account signature type from a single signature at credential
+    /// and key indices 0.
+    pub fn singleton(signature: Signature) -> Self {
+        let credential_map = CredentialSignatures {
+            sigs: [(0, signature)].into_iter().collect(),
+        };
+
+        AccountSignatures {
+            sigs: [(0, credential_map)].into_iter().collect(),
+        }
+    }
+}
+
+/// Account signatures verification data includes the [`Signature`]s and
+/// associated [`VerifyKey`]s for some credentials and key indices.
+///
+/// Account signatures verification data should be thought of as a nested map,
+/// indexed on the outer layer by credential indices, and the inner map maps key
+/// indices to [`Signature`]s and [`VerifyKey`]s.
+#[derive(Debug, Clone, PartialEq, Eq, serde::Deserialize, serde::Serialize)]
+#[repr(transparent)]
+pub struct AccountSignaturesVerificationData {
+    pub data: BTreeMap<u8, CredentialSignaturesVerificationData>,
+}
+
+#[derive(Debug, Clone, PartialEq, Eq, serde::Deserialize, serde::Serialize)]
+#[repr(transparent)]
+pub struct CredentialSignaturesVerificationData {
+    pub data: BTreeMap<u8, AccountSignaturesVerificationEntry>,
+}
+
+#[derive(Debug, Clone, PartialEq, Eq, serde::Deserialize, serde::Serialize)]
+pub struct AccountSignaturesVerificationEntry {
+    pub signature:  Signature,
+    pub public_key: VerifyKey,
+}
+
+impl AccountSignaturesVerificationData {
+    /// Create the account signatures verification data type from a single
+    /// signature and public key at credential and key indices 0.
+    pub fn singleton(signature: Signature, public_key: VerifyKey) -> Self {
+        let credential_map = CredentialSignaturesVerificationData {
+            data: [(0, AccountSignaturesVerificationEntry {
+                signature,
+                public_key,
+            })]
+            .into_iter()
+            .collect(),
+        };
+
+        AccountSignaturesVerificationData {
+            data: [(0, credential_map)].into_iter().collect(),
+        }
+    }
+
+    /// Zip an [`AccountSignatures`] and an [`AccountKeys`] into
+    /// an [`AccountSignaturesVerificationData`] type. The function reverts if
+    /// the indices in the maps do not match.
+    pub fn zip_signatures_and_keys(
+        account_signatures: &AccountSignatures,
+        account_keys: &AccountKeys,
```

Comment:
To avoid cloning, we would also need to change the return parameter signature and this doesn't seem much simpler.
```
struct AccountSignaturesVerificationEntry<'a> {
    signature: &'a Signature,
    public_key: &'a PublicKey,
}
```

Reply:
It is possible without changing the return type, you would also need to change some references inside the function:
```rust
pub fn zip_signatures_and_keys(
        account_signatures: AccountSignatures,
        account_keys: AccountKeys,
    ) -> Result<Self, SignatureError> {
        let mut outer_map = BTreeMap::new();

        // Check that the length of the outer map matches.
        if account_signatures.sigs.len() != account_keys.keys.len() {
            return Err(SignatureError::MismatchMapIndices);
        }

        for (outer_key, credential_sigs) in account_signatures.sigs {
            // Check if corresponding account key exists.
            if let Some(account_key_pair) =
                account_keys.keys.get(&CredentialIndex { index: outer_key })
            {
                let public_keys = account_key_pair.get_public_keys();

                // Check that the length of the inner map matches.
                if credential_sigs.sigs.len() != public_keys.len() {
                    return Err(SignatureError::MismatchMapIndices);
                }

                // Create the zipped inner map.
                let inner_map: Result<
                    BTreeMap<u8, AccountSignaturesVerificationEntry>,
                    SignatureError,
                > = credential_sigs
                    .sigs
                    .into_iter()
                    .zip(public_keys.into_iter())
                    .map(|((inner_key, signature), (key_index, public_key))| {
                        // Ensure that inner_key and key_index matches.
                        if inner_key != key_index.0 {
                            return Err(SignatureError::MismatchMapIndices);
                        }
                        Ok((inner_key, AccountSignaturesVerificationEntry {
                            signature,
                            public_key,
                        }))
                    })
                    .collect();

                outer_map.insert(outer_key, CredentialSignaturesVerificationData {
                    data: inner_map?,
                });
            } else {
                return Err(SignatureError::MismatchMapIndices);
            }
        }

        Ok(AccountSignaturesVerificationData { data: outer_map })
    }
}
```